### PR TITLE
[SPARK-40387][SQL] Improve the implementation of Spark Decimal

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR used to improve the implementation of Spark `Decimal`. The improvement points are as follows:
1. Use `toJavaBigDecimal` instead of  `toBigDecimal.bigDecimal` 
2. Extract `longVal / POW_10(_scale)` as a new method `def actualLongVal: Long`
3. Remove `BIG_DEC_ZERO` and use `decimalVal.signum` to judge whether or not equals zero.
4. Use `<` instead of `compare`.
5. Correct some code style. 

### Why are the changes needed?
Improve the implementation of Spark Decimal

### Does this PR introduce _any_ user-facing change?
'No'.
Just update the inner implementation.

### How was this patch tested?
N/A

### CODE REVIEW

**Comment by beliefer** [comment]
The failure test case is not related to this PR.
@MaxGekk @gengliangwang cc @cloud-fan

**Comment by cloud-fan** [comment]
I don't think `eq` and `ne` has observable perf difference, can we revert this part?

**Comment by beliefer** [comment]
> I don't think `eq` and `ne` has observable perf difference, can we revert this part?

OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -184,68 +184,56 @@ final class Decimal extends Ordered[Decimal] with Serializable {
     this
   }
 
-  def toBigDecimal: BigDecimal = {
-    if (decimalVal.ne(null)) {
-      decimalVal
-    } else {
-      BigDecimal(longVal, _scale)
-    }
+  def toBigDecimal: BigDecimal = if (decimalVal.ne(null)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:187]: can we avoid pure code style change? the previous code style was not wrong.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:187]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 2]
```diff
@@ -240,9 +240,11 @@ final class Decimal extends Ordered[Decimal] with Serializable {
 
   def toFloat: Float = toBigDecimal.floatValue
 
+  private def actualLongVal: Long = longVal / POW_10(_scale)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:243]: We need a better name to differentiate this from `toLong`. How about `rawLongValue`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 3]
```diff
@@ -487,12 +488,12 @@ final class Decimal extends Ordered[Decimal] with Serializable {
       DecimalType.MAX_SCALE, MATH_CONTEXT.getRoundingMode))
 
   def % (that: Decimal): Decimal =
-    if (that.isZero) null
-    else Decimal(toJavaBigDecimal.remainder(that.toJavaBigDecimal, MATH_CONTEXT))
+    if (that.isZero) null else Decimal(toJavaBigDecimal.remainder(that.toJavaBigDecimal,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:491]: let's avoid pure code style changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 4]
```diff
@@ -504,7 +505,7 @@ final class Decimal extends Ordered[Decimal] with Serializable {
     }
   }
 
-  def abs: Decimal = if (this.compare(Decimal.ZERO) < 0) this.unary_- else this
+  def abs: Decimal = if (this < Decimal.ZERO) this.unary_- else this
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:508]: does this have a real impact?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:508]: `this < Decimal.ZERO` is more clear.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 5]
```diff
@@ -240,9 +240,11 @@ final class Decimal extends Ordered[Decimal] with Serializable {
 
   def toFloat: Float = toBigDecimal.floatValue
 
+  private def rawLongValue: Long = longVal / POW_10(_scale)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:243]: let's still use the old name `actualLongVal`

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 5 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: OK

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thank you !

**Pull request closed without merging**